src/audio/tdfb/tdfb.c

```diff
@@ -148,10 +218,102 @@ static int tdfb_init_coef(struct tdfb_comp_data *cd, int source_nch,
 		return -EINVAL;
 	}
 
-	coefp = ASSUME_ALIGNED(&config->data[0], 2);
+	if (config->num_angles > SOF_TDFB_MAX_ANGLES) {
+		comp_cl_err(&comp_tdfb, "tdfb_init_coef(), invalid num_angles %d",
+			    config->num_angles);
+		return -EINVAL;
+	}
+
+	if (config->beam_off_defined > 1) {
+		comp_cl_err(&comp_tdfb, "tdfb_init_coef(), invalid beam_off_defined %d",
+			    config->beam_off_defined);
+		return -EINVAL;
+	}
+
+	if (config->num_mic_locations > SOF_TDFB_MAX_MICROPHONES) {
+		comp_cl_err(&comp_tdfb, "tdfb_init_coef(), invalid num_mic_locations %d",
+			    config->num_mic_locations);
+		return -EINVAL;
+	}
+
+	/* In SOF v1.6 - 1.8 based beamformer topologies the multiple angles, mic locations,
+	 * and beam on/off switch were not defined. Return error if such configuration is seen.
+	 * A most basic blob has num_angles equals 1. Mic locations data is optional.
+	 */
+	if (config->num_angles == 0 && config->num_mic_locations == 0) {
+		comp_cl_err(&comp_tdfb, "tdfb_init_coef(), ABI version less than 3.19.1 is not supported.");
+		return -EINVAL;
+	}
+
+	/* Skip filter coefficients */
+	num_filters = config->num_filters * (config->num_angles + config->beam_off_defined);
+	coefp = tdfb_filter_seek(config, num_filters);
+
+	/* Get shortcuts to input and output configuration */
+	cd->input_channel_select = coefp;
+	coefp += config->num_filters;
+	cd->output_channel_mix = coefp;
+	coefp += config->num_filters;
+	cd->output_stream_mix = coefp;
+	coefp += config->num_filters;
+
+	/* Check if there's beam-off configured, then get pointers to beam angles data
+	 * and microphone locations. Finally check that size matches.
+	 */
+	if (config->beam_off_defined) {
+		output_channel_mix_beam_off = coefp;
+		coefp += config->num_filters;
+	}
+	cd->filter_angles = (struct sof_tdfb_angle *)coefp;
+	cd->mic_locations = (struct sof_tdfb_mic_location *)
+		(&cd->filter_angles[config->num_angles]);
+	if ((char *)&cd->mic_locations[config->num_mic_locations] !=
+	    (char *)config + config->size) {
+		comp_cl_err(&comp_tdfb, "tdfb_init_coef(), invalid config size");
+		return -EINVAL;
+	}
+
+	/* Skip to requested coefficient set */
+	min_delta = 360;
+	min_delta_idx = 0;
+	target_az = cd->az_value * config->angle_enum_mult + config->angle_enum_offs;
+	if (target_az > 180)
+		target_az -= 360;
+
+	if (target_az < -180)
+		target_az += 360;
+
+	for (i = 0; i < config->num_angles; i++) {
+		delta = ABS(target_az - cd->filter_angles[i].azimuth);
+		if (delta < min_delta) {
+			min_delta = delta;
+			min_delta_idx = i;
+		}
+	}
+
+	idx = cd->filter_angles[min_delta_idx].filter_index;
+	if (cd->beam_on) {
+		comp_cl_info(&comp_tdfb, "tdfb_init_coef(), angle request %d, found %d, idx %d",
+			     target_az, cd->filter_angles[min_delta_idx].azimuth, idx);
+	} else {
+		if (config->beam_off_defined) {
```

Comment:
could do this as `else if (config->beam_off_defined) {`

Reply:
Done